source/s3_auto_ranged_put.c

```diff
@@ -36,6 +38,196 @@ static void s_s3_auto_ranged_put_request_finished(
+    }
+
+    if (!aws_byte_cursor_eq_c_str(&type_value, "AWS_S3_META_REQUEST_TYPE_PUT_OBJECT")) {
+        AWS_LOGF_ERROR(AWS_LS_S3_META_REQUEST, "Could not load persisted state. Invalid token type.");
+        goto invalid_argument_cleanup;
+    }
+
+    if ((size_t)partition_size_value < g_s3_min_upload_part_size) {
+        AWS_LOGF_ERROR(
+            AWS_LS_S3_META_REQUEST,
+            "Could not create resume auto-ranged-put meta request; part size of %" PRIu64
+            " specified in the token is below minimum threshold for multi-part.",
+            (uint64_t)partition_size_value);
+
+        goto invalid_argument_cleanup;
+    }
+
+    if ((uint32_t)total_num_parts_value > g_s3_max_num_upload_parts) {
+        AWS_LOGF_ERROR(
+            AWS_LS_S3_META_REQUEST,
+            "Could not create resume auto-ranged-put meta request; total number of parts %" PRIu32
+            " specified in the token is too large for platform.",
+            (uint32_t)total_num_parts_value);
+
+        goto invalid_argument_cleanup;
+    }
+
+    *upload_id_out = aws_string_new_from_cursor(allocator, &multipart_upload_id_value);
+    *part_size_out = (size_t)partition_size_value;
+    *total_num_parts_out = (uint32_t)total_num_parts_value;
+
+    aws_json_value_destroy(root);
+    aws_json_module_cleanup();
+
+    return AWS_OP_SUCCESS;
+
+invalid_argument_cleanup:
+    aws_json_value_destroy(root);
+    aws_json_module_cleanup();
+    return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+}
+
+/**
+ * Loads the persistable state used to resume an upload that was previously paused.
+ */
+static int s_load_persistable_state(
+    struct aws_allocator *allocator,
+    struct aws_s3_auto_ranged_put *auto_ranged_put,
+    uint64_t content_length,
+    struct aws_string *upload_id,
+    size_t part_size,
+    uint32_t total_num_parts) {
+
+    uint32_t num_parts = (uint32_t)(content_length / part_size);
+
+    if ((content_length % part_size) > 0) {
+        ++num_parts;
+    }
+
+    if (total_num_parts != num_parts) {
+        AWS_LOGF_ERROR(
+            AWS_LS_S3_META_REQUEST,
+            "Could not create auto-ranged-put meta request; persisted number of parts %" PRIu32
+            " does not match expected number of parts based on length of the body.",
+            total_num_parts);
+
+        return AWS_ERROR_INVALID_ARGUMENT;
+    }
+
+    auto_ranged_put->synced_data.num_parts_sent = 0;
+    auto_ranged_put->synced_data.num_parts_completed = 0;
+    auto_ranged_put->synced_data.create_multipart_upload_sent = true;
+    auto_ranged_put->synced_data.create_multipart_upload_completed = true;
+    auto_ranged_put->upload_id = upload_id;
+
+    struct aws_byte_cursor request_path;
+    if (aws_http_message_get_request_path(auto_ranged_put->base.initial_request_message, &request_path)) {
+        AWS_LOGF_ERROR(AWS_LS_S3_META_REQUEST, "Could not load persisted state. Request path could not be read.");
+        return AWS_ERROR_INVALID_ARGUMENT;
+    }
```

Comment:
trivial: can we move this a bit to line 187? Feel more natural if all those argument check stay together. Also, `return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);`